### PR TITLE
fix timespec 2038 overflow in kernel >= 5

### PR DESCRIPTION
**Pull request by moonbuggy**
I previously patched this to handle deprecated time.h functions in kernels >=5.0. It turns out my fix is itself deprecated by kernel 5.6, so here we are again. :)

We're now storing TES_P2P_0002_tmval as a 64-bit value and using ktime_get_real_ts64() instead of getnstimeofday(). This fixes an overflow in 2038 and allows building against the latest kernels.

As with my previous effort, this is only happening for builds against kernel 5.0 or newer, although it should work as far back as [3.17](https://elixir.bootlin.com/linux/v3.17/ident/ktime_get_real_ts64). I didn't want to make changes beyond what I'm building and testing, so I'll leave it to you to decide if you want to apply this patch any further back.

The only practical difference between this and the existing code for older versions of the kernel is, afaik, that they'll stop working in 18 years.

### CODE REVIEW

**Comment by fifteenhex** [comment]
I don't have anything to test this right now but I'll trust it works. 

Out of interest what are you using this driver on and does it actually work? Last time I checked with the orangepi zero so many frames were dropped it wasn't much use but maybe that was actually a hardware issue and it works well for you?

**Comment by moonbuggy** [comment]
The radio usually (not always) comes up when I bring it up and I've moved data through it. I haven't really tested it properly myself, there may be a bunch of frames going missing.

This is a [project](https://github.com/moonbuggy/orange-pi-zero-alpine) I keep getting distracted from, so it's not really where I want it to be yet. Too much starting and stopping when other things come up. As a result I'm still poking around in a lot of different areas at once,

I did have it running as an access point with hostapd and wpa-supplicant at various points last year (kernels 5.0-5.3), I believe. I don't recall how much traffic I moved through it though. I've only run it (briefly) as a normal wireless client in the current build (5.6rc4) and it did sometimes fail to create a link, but I didn't put the time in to troubleshoot it and there were factors other than xradio in play. It was simpler to use a cable than to worry about why I didn't always have WiFi at boot, I was focused on the I2C bus at the time. :)

I do intend to look at the WiFi more thoroughly at some point. No idea when I'll get to that point though. :) So far I've really just been making sure that it builds and turns on, so that it's available when I do get a chance to look properly.

TL;DR: It _seems_ to generally work, I don't yet know if it works always or perfectly. :)

**Comment by techcap** [comment]
On my experience, kernel panic is occurred if wifi channel is changed by router or access point.

**Comment by moonbuggy** [comment]
I got a bit curious, so I just had a quick but closer look at the WiFi. There's a problem with the interrupts, but that's the only issue that stands out.

I changed the WiFi channel on the router while connected. dmesg looks like this:

```
[  646.064573] wlan0: Connection to AP 14:dd:a9:d0:ab:70 lost
[  646.154706] xradio_wlan mmc1:0001:1: missed interrupt
[  646.194637] xradio_wlan mmc1:0001:1: missed interrupt
[  646.195401] ieee80211 phy0: no station to remove
[  646.196230] ieee80211 phy0: vif 0, configuring tx
[  646.196667] ieee80211 phy0: vif 0, configuring tx
[  646.197025] ieee80211 phy0: vif 0, configuring tx
[  646.197367] ieee80211 phy0: vif 0, configuring tx
[  646.197861] ieee80211 phy0: ignore IEEE80211_CONF_CHANGE_MONITOR (0)IEEE80211_CONF_CHANGE_IDLE (1)
[  646.334607] xradio_wlan mmc1:0001:1: missed interrupt
[  646.654590] xradio_wlan mmc1:0001:1: missed interrupt
[  646.694586] xradio_wlan mmc1:0001:1: missed interrupt
[  646.734595] xradio_wlan mmc1:0001:1: missed interrupt
[  646.774577] xradio_wlan mmc1:0001:1: missed interrupt
[  647.094595] xradio_wlan mmc1:0001:1: missed interrupt
[  647.096161] wlan0: authenticate with 14:dd:a9:d0:ab:70
[  647.096233] ieee80211 phy0: ignore IEEE80211_CONF_CHANGE_MONITOR (0)IEEE80211_CONF_CHANGE_IDLE (1)
[  647.096243] ieee80211 phy0: Freq 2457 (wsm ch: 10).
[  647.096459] wlan0: send auth to 14:dd:a9:d0:ab:70 (try 1/3)
[  647.134594] xradio_wlan mmc1:0001:1: missed interrupt
[  647.134683] xradio_wlan mmc1:0001:1: received frame has no key status
[  647.134769] xradio_wlan mmc1:0001:1: received frame has no key status
[  647.174599] xradio_wlan mmc1:0001:1: missed interrupt
[  647.214598] xradio_wlan mmc1:0001:1: missed interrupt
[  647.214771] wlan0: authenticated
[  647.224578] wlan0: associate with 14:dd:a9:d0:ab:70 (try 1/3)
[  647.264612] xradio_wlan mmc1:0001:1: missed interrupt
[  647.264827] wlan0: RX AssocResp from 14:dd:a9:d0:ab:70 (capab=0x411 status=0 aid=2)
[  647.264902] ieee80211 phy0: vif 0, configuring tx
[  647.265748] ieee80211 phy0: vif 0, configuring tx
[  647.266077] ieee80211 phy0: vif 0, configuring tx
[  647.266398] ieee80211 phy0: vif 0, configuring tx
[  647.304602] xradio_wlan mmc1:0001:1: missed interrupt
[  647.305891] wlan0: associated
```

The "missed interrupt" lines are constantly occurring, not just during the channel hop. No kernel panic.

iperf3 looks like this:
```
Connecting to host 192.168.30.132, port 5201
[  5] local 192.168.30.50 port 39054 connected to 192.168.30.132 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  12.8 MBytes  12.8 MBytes/sec   51    215 KBytes
[  5]   1.00-2.00   sec  11.0 MBytes  11.0 MBytes/sec    0    252 KBytes
[  5]   2.00-3.00   sec  11.6 MBytes  11.6 MBytes/sec   48    202 KBytes
[  5]   3.00-4.00   sec  11.2 MBytes  11.1 MBytes/sec    0    242 KBytes
[  5]   4.00-5.00   sec  11.6 MBytes  11.6 MBytes/sec    0    275 KBytes
[  5]   5.00-6.00   sec  11.0 MBytes  11.0 MBytes/sec   42    231 KBytes
[  5]   6.00-7.00   sec  11.6 MBytes  11.6 MBytes/sec    0    265 KBytes
[  5]   7.00-8.00   sec  11.2 MBytes  11.1 MBytes/sec   48    220 KBytes
[  5]   8.00-9.00   sec  11.0 MBytes  11.0 MBytes/sec    0    255 KBytes
[  5]   9.00-10.00  sec  11.6 MBytes  11.7 MBytes/sec   40    207 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec   115 MBytes  11.5 MBytes/sec  229             sender
[  5]   0.00-10.05  sec   113 MBytes  11.3 MBytes/sec                  receiver

iperf Done.
```

The high retransmission count is presumably related to the interrupt problem. I ran iperf a few times, either side of some channel hops, and the output is consistent between runs.

The device tree physically on the OPiZero isn't stable at the moment though, I've been playing with it sort of generally lately. It's possible I've introduced an error there that's causing the missed interrupts. I just haven't look at it in detail, so don't know.

I don't remember if the interrupts were working better in older builds. I'll probably have a look at that at some point, when I get around to properly troubleshooting the WiFi. It still isn't my top priority though, so it might be a while before I get to that. 

I can give you an update if/when I make some progress in that area, if you're interested.

**Comment by fifteenhex** [comment]
Interesting. It doesn't seem too bad for you. I think interrupts have always been bad. I don't think that's on the xradio side but just an issue with the latency of the SoC interrupt they used.

I might get my orange pi zero out over the weekend and take a look.

**Comment by sanjuruk** [comment]
When I tested xradio_wlan driver resulted in Kernel hang.

Please refer this [Armbian thread](https://forum.armbian.com/topic/10493-opi-zero-xradio_wlan-driver-kernel-hang/) about my problems and my dead end.

Just to summarize,
I noticed problems with xradio in 2 modes.
1) When using xradio in AP mode: kernel hang, reboot -f does not work
2) Using xradio in concurrent mode: kernel hang occurred

In either cases AP mode had the problem.

When xradio(in AP) changed the channel, I noticed Kernal hang occurs
```
[20148.872296] rcu: INFO: rcu_sched self-detected stall on CPU
[20148.877893] rcu:     1-....: (4945709 ticks this GP) idle=8aa/1/0x40000002 softirq=9640/9640 fqs=2447771
[20148.887182] rcu:      (t=4967445 jiffies g=15521 q=274313)
[20148.892407] NMI backtrace for cpu 1
[20148.895897] CPU: 1 PID: 396 Comm: xradio_bh Tainted: G      D W         4.19.20-sunxi #5.75
[20148.904234] Hardware name: Allwinner sun8i Family
[20148.908959] [<c010dbbd>] (unwind_backtrace) from [<c010a7b1>] (show_stack+0x11/0x14)
[20148.916701] [<c010a7b1>] (show_stack) from [<c08c8dc1>] (dump_stack+0x69/0x78)
[20148.923921] [<c08c8dc1>] (dump_stack) from [<c08cd03b>] (nmi_cpu_backtrace+0x8f/0x90)
[20148.931746] [<c08cd03b>] (nmi_cpu_backtrace) from [<c08cd0eb>] (nmi_trigger_cpumask_backtrace+0xaf/0xe0)
[20148.941216] [<c08cd0eb>] (nmi_trigger_cpumask_backtrace) from [<c016c183>] (rcu_dump_cpu_stacks+0x7b/0x98)
[20148.950861] [<c016c183>] (rcu_dump_cpu_stacks) from [<c016b7b1>] (rcu_check_callbacks+0x4f5/0x6c8)
[20148.959815] [<c016b7b1>] (rcu_check_callbacks) from [<c0170af3>] (update_process_times+0x2b/0x48)
[20148.968681] [<c0170af3>] (update_process_times) from [<c017e96b>] (tick_sched_timer+0x37/0x74)
[20148.977277] [<c017e96b>] (tick_sched_timer) from [<c0171355>] (__hrtimer_run_queues+0x105/0x254)
[20148.986055] [<c0171355>] (__hrtimer_run_queues) from [<c0171e15>] (hrtimer_interrupt+0xb5/0x200)
[20148.994835] [<c0171e15>] (hrtimer_interrupt) from [<c0780ab1>] (arch_timer_handler_phys+0x25/0x28)
[20149.003787] [<c0780ab1>] (arch_timer_handler_phys) from [<c0162d9b>] (handle_percpu_devid_irq+0x57/0x19c)
[20149.013346] [<c0162d9b>] (handle_percpu_devid_irq) from [<c015f1f5>] (generic_handle_irq+0x1d/0x28)
[20149.022383] [<c015f1f5>] (generic_handle_irq) from [<c015f695>] (__handle_domain_irq+0x45/0x84)
[20149.031076] [<c015f695>] (__handle_domain_irq) from [<c059a475>] (gic_handle_irq+0x39/0x68)
[20149.039421] [<c059a475>] (gic_handle_irq) from [<c0101a65>] (__irq_svc+0x65/0x94)
[20149.046892] Exception stack(0xd66c9de8 to 0xd66c9e30)
[20149.051941] 9de0:                   c9ae0e4c 00000000 0000287c 0000287b 00000000 d6480e00
[20149.060110] 9e00: c0d04d48 c9ae0d18 d648100c 00000080 d64811fc 00000000 ce9bc61c d66c9e38
[20149.068276] 9e20: bfb2d42d c08db396 20070033 ffffffff
[20149.073328] [<c0101a65>] (__irq_svc) from [<c08db396>] (_raw_spin_lock+0x26/0x34)
[20149.080830] [<c08db396>] (_raw_spin_lock) from [<bfb2d42d>] (wsm_handle_rx+0x828/0xc7c [xradio_wlan])
[20149.090066] [<bfb2d42d>] (wsm_handle_rx [xradio_wlan]) from [<bfb29e05>] (xradio_bh_exchange+0x27c/0x588 [xradio_wlan])
[20149.100862] [<bfb29e05>] (xradio_bh_exchange [xradio_wlan]) from [<bfb2a239>] (xradio_bh+0x128/0x270 [xradio_wlan])
[20149.111301] [<bfb2a239>] (xradio_bh [xradio_wlan]) from [<c0132cb1>] (kthread+0xfd/0x104)
[20149.119470] [<c0132cb1>] (kthread) from [<c01010f9>] (ret_from_fork+0x11/0x38)
[20149.126680] Exception stack(0xd66c9fb0 to 0xd66c9ff8)
[20149.131726] 9fa0:                                     00000000 00000000 00000000 00000000
[20149.139894] 9fc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[20149.148061] 9fe0: 00000000 00000000 00000000 00000000 00000013 00000000
```
AP is managed by NetworkManager, and also used by hostapd and dnsmaq
```
wlan0_ap: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.10.10.1  netmask 255.255.255.0  broadcast 10.10.10.255
        inet6 fe80::1042:93ff:fe7e:5071  prefixlen 64  scopeid 0x20<link>
        ether 12:42:93:7e:50:71  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 69  bytes 11042 (11.0 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

After checking the error message ```[20149.090066] [<bfb2d42d>] (wsm_handle_rx﻿ [xradio_wlan]) from [<bfb29e05>] (xradio_bh_exchange+0x27c/0x588 [xradio_wlan])```, it seems program stalled on wsm_handle_rx in xradio_wlan.

At [wsm.c 1780 line](https://github.com/fifteenhex/xradio/blob/8965ea55ddfd862ee76384913f82e6b69628cd27/wsm.c#L1780) 
```
if (raceCheck == 0xFFFF) {
			/* If wsm_handle_rx got stuck in _confirm we will hang
			 * system there. It's better than silently currupt
			 * stack or heap, isn't it? */
			BUG_ON(wait_event_timeout(
					hw_priv->wsm_cmd_wq,
					hw_priv->wsm_cmd.done,
					WSM_CMD_LAST_CHANCE_TIMEOUT) <= 0);
		}
```
it seems like the kernel hang is done on purpose ?! 
Is there a way to recover?

**Comment by fifteenhex** [comment]
The BUG_ON will cause a dead lock. The idea behind BUG_ON is that it is triggered on something that would break your data etc hence it tries to stop the system from doing anything further.
I would try taking the BUG_ON out and replacing it with some code that tries to check if the action has happened but the interrupt was missed.
